File: server/shared/typescript/validation/default/src/default-postal-code-validator.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import type { CountryCode } from "@d2/geo-abstractions";
import { TK } from "@d2/i18n/keys";
import {
  inputError,
  ok,
  tk,
  validationFailed,
  type D2Result,
} from "@d2/result";
import { falsey } from "@d2/utilities";
import type { IPostalCodeValidator } from "@d2/validation-abstractions";
import { postcodeValidator } from "postcode-validator";

/**
 * Default `IPostalCodeValidator` implementation. Mirrors the .NET
 * `D2.Shared.Validation.DefaultPostalCodeValidator` — both run a
 * country-aware structural check and normalize to trimmed + uppercased form.
 * An unknown country fails closed (ValidationFailed, never a throw) on both
 * runtimes, so cross-language behavior on unsupported countries is identical.
 */
export class DefaultPostalCodeValidator implements IPostalCodeValidator {
  /**
   * Validates the supplied postal code and returns its normalized form on
   * success.
   *
   * @param postalCode - The postal code to validate (may be `undefined`,
   *   empty, or whitespace).
   * @param countryCode - Optional country whose postal-code format governs
   *   validation. An unknown / unsupported country fails closed.
   * @returns `ok` wrapping the trimmed + uppercased postal code on success;
   *   `validationFailed` keyed `"postalCode"` with `POSTAL_CODE_INVALID` on
   *   `undefined`, empty, whitespace, structurally invalid, or
   *   unknown-country input.
   */
  validate(
    postalCode: string | undefined,
    countryCode?: CountryCode,
  ): D2Result<string> {
    if (falsey(postalCode)) return DefaultPostalCodeValidator.invalid();

    // Fail closed when no country is supplied — an absent country cannot
    // select a per-country format, so there is nothing to validate against.
    // This mirrors the .NET `DefaultPostalCodeValidator`, which returns
    // `ValidationFailed` for a null country. There is deliberately NO
    // permissive country-agnostic fallback (e.g. the `"INTL"` pattern): an
    // unknown / absent country always fails closed on both runtimes.
    if (countryCode === undefined) return DefaultPostalCodeValidator.invalid();

    // `falsey` already excluded undefined / empty / whitespace.
    // Normalize (trim + uppercase) before validating so the regex sees the
    // same form the .NET side matches against. `postcode-validator` applies
    // the per-country regex verbatim, so the uppercase normalization here is
    // what keeps case-insensitive inputs (e.g. "k1a 0b1") matching.
    const normalized = postalCode!.trim().toUpperCase();

    // The `CountryCode` brand erases to its underlying alpha-2 string at
    // runtime, the form `postcode-validator` keys its per-country regexes on.
    const country = countryCode as string;

    // Fail closed: `postcodeValidator` throws for an unknown country code.
    // Treat any throw as a validation failure rather than letting it escape.
    let isValid: boolean;
    try {
      isValid = postcodeValidator(normalized, country);
    } catch {
      return DefaultPostalCodeValidator.invalid();
    }

    if (!isValid) return DefaultPostalCodeValidator.invalid();

    return ok<string>(normalized);
  }

  private static invalid(): D2Result<string> {
    return validationFailed<string>({
      inputErrors: [
        inputError("postalCode", [
          tk(TK.common.validation.POSTAL_CODE_INVALID),
        ]),
      ],
    });
  }
}

File: public/packages/dotnet/validation/default/DefaultPostalCodeValidator.cs
// -----------------------------------------------------------------------
// <copyright file="DefaultPostalCodeValidator.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Validation;

using System.Text.RegularExpressions;
using DcsvIo.D2.Geo.Abstractions;
using DcsvIo.D2.I18n;
using DcsvIo.D2.Result;
using DcsvIo.D2.Utilities.Extensions;
using DcsvIo.D2.Validation.Abstractions;

/// <summary>
/// Default <see cref="IPostalCodeValidator"/> — country-aware structural check
/// using a per-country regex map ported from <c>postcode-validator@3.10.9</c>.
/// Mirrors the TypeScript <c>@dcsv-io/d2-validation</c> <c>DefaultPostalCodeValidator</c>
/// in contract and behavior.
/// </summary>
/// <remarks>
/// <para>
/// <b>Fail-closed policy.</b> When the country code is <see langword="null"/> or
/// its alpha-2 key is absent from the dataset, validation fails with
/// <see cref="TK.Common.Validation.POSTAL_CODE_INVALID"/>. There is no fallback
/// to a permissive global-range pattern — unknown-country input always fails closed.
/// </para>
/// <para>
/// <b>Normalization.</b> The input is trimmed and uppercased before the regex
/// runs and before the normalized value is returned.
/// </para>
/// <para>
/// <b>ReDoS posture: anchored, bounded — no catastrophic backtracking
/// (Bucket B2 + defense-in-depth).</b>
/// All per-country patterns are anchored at both ends, compiled, and carry a 50 ms
/// <c>matchTimeout</c>. The regexes are sourced verbatim from
/// <c>postcode-validator@3.10.9</c> (patterns that lacked anchors in the
/// source were wrapped as <c>^(?:...)$</c>). Because a future dataset update
/// could introduce a pattern with super-linear backtracking, the 50 ms timeout
/// is retained as a mandatory guard-rail on every entry. The map is built once
/// at static initialization via <see cref="PostalCodeRegexData.SR_Map"/>.
/// </para>
/// </remarks>
public sealed class DefaultPostalCodeValidator : IPostalCodeValidator
{
    static DefaultPostalCodeValidator()
    {
        // Pre-warm the static map so the first real validation call does not
        // incur the JSON deserialization + FrozenDictionary build cost.
        // PostalCodeRegexData.SR_Map access triggers the type initializer.
        _ = PostalCodeRegexData.SR_Map.Count;
    }

    /// <inheritdoc />
    public D2Result<string> Validate(string? postalCode, CountryCode? countryCode = null)
    {
        if (postalCode.Falsey())
            return Invalid();

        var normalized = postalCode!.Trim().ToUpperInvariant();

        if (countryCode is null)
            return Invalid();

        // CountryCode enum names are the ISO 3166-1 alpha-2 codes (e.g. US, GB).
        var key = countryCode.Value.ToString();

        if (!PostalCodeRegexData.SR_Map.TryGetValue(key, out var regex))
            return Invalid();

        bool isMatch;
        try
        {
            isMatch = regex.IsMatch(normalized);
        }
        catch (RegexMatchTimeoutException)
        {
            return Invalid();
        }

        if (!isMatch)
            return Invalid();

        return D2Result<string>.Ok(normalized);
    }

    private static D2Result<string> Invalid()
        => D2Result<string>.ValidationFailed(
            inputErrors:
            [
                new InputError("postalCode", [TK.Common.Validation.POSTAL_CODE_INVALID]),
            ]);
}
